src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -526,26 +525,69 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
         /// deepCheckTupleNames should be set to false when the caller is already visiting constituent types.
         /// </summary>
-        private static bool CheckTupleNamesConstraints(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder, bool deepCheckTupleNames)
+        private static bool CheckTupleNamesConstraints(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder, TupleNamesCheckKind deepCheckTupleNames)
         {
-            if (deepCheckTupleNames)
+            if (!type.ContainsTuple())
             {
-                object failedType = type.VisitType((t, diag, _) => !CheckTupleNamesConstraints(t, diag, deepCheckTupleNames: false), builder);
-                return failedType == null;
+                return true;
             }
-            else
+
+            switch (deepCheckTupleNames)
             {
-                if (!CheckTupleNamesConstraintsCore(type, builder))
-                {
-                    return false;
-                }
+                case TupleNamesCheckKind.Shallow:
+                    {
+                        if (!CheckTupleNamesConstraintsCore(type, builder))
+                        {
+                            return false;
+                        }
 
-                var baseType = type.BaseTypeNoUseSiteDiagnostics;
-                return (object)baseType != null ? CheckTupleNamesConstraints(baseType, builder, deepCheckTupleNames: false) : true;
+                        // Walk up the base types and apply the same check on each inheritence level.
+                        // If the base type was not "variable", then we can cut short.
+                        TypeSymbol currentType = type;
+                        NamedTypeSymbol baseType = currentType.BaseTypeNoUseSiteDiagnostics;
+                        bool result = true;
+                        while ((object)baseType != null && currentType.OriginalDefinition.BaseTypeNoUseSiteDiagnostics.ContainsTypeParameter())
+                        {
+                            result &= CheckTupleNamesConstraints(baseType, builder, TupleNamesCheckKind.Shallow);
+                            currentType = baseType;
+                            baseType = currentType.BaseTypeNoUseSiteDiagnostics;
+                        }
+                        return result;
+                    }
+                case TupleNamesCheckKind.DeepExceptTopLevel:
+                    {
+                        if (!CheckTupleNamesConstraintsCore(type, builder))
+                        {
+                            return false;
+                        }
+                        NamedTypeSymbol baseType = type.BaseTypeNoUseSiteDiagnostics;
+                        // If the base type was not "variable", then we can cut short.
+                        if ((object)baseType == null || !type.OriginalDefinition.BaseTypeNoUseSiteDiagnostics.ContainsTypeParameter())
```

Comment:
I see, thanks. Perhaps add a comment that any errors on a non-generic base type should have been reported on the base type itself.